nestkernel/simulation_manager.cpp

```diff
@@ -375,6 +376,61 @@ nest::SimulationManager::simulate( Time const& t )
   t_slice_begin_ = timeval();
   t_slice_end_ = timeval();
 
+  // Reset profiling timers and counters within event_delivery_manager
+  kernel().event_delivery_manager.reset_timers_counters();
+
+  // find shortest and longest delay across all MPI processes
+  // this call sets the member variables
+  kernel().connection_manager.update_delay_extrema_();
+  kernel().event_delivery_manager.init_moduli();
+
+  // Check for synchronicity of global rngs over processes.
+  // We need to do this ahead of any simulation in case random numbers
+  // have been consumed on the SLI level.
+  if ( kernel().mpi_manager.get_num_processes() > 1 )
+  {
+    if ( !kernel().mpi_manager.grng_synchrony(
+           kernel().rng_manager.get_grng()->ulrand( 100000 ) ) )
+    {
+      LOG( M_ERROR,
+        "SimulationManager::simulate",
+        "Global Random Number Generators are not synchronized prior to "
+        "simulation." );
+      throw KernelException();
+    }
+  }
+
+  // if at the beginning of a simulation, set up spike buffers
+  if ( !simulated_ )
+    kernel().event_delivery_manager.configure_spike_buffers();
+
+  kernel().node_manager.ensure_valid_thread_local_ids();
+  num_active_nodes = kernel().node_manager.prepare_nodes();
+
+  kernel().model_manager.create_secondary_events_prototypes();
+
+  // we have to do enter_runtime after prepre_nodes, since we use
+  // calibrate to map the ports of MUSIC devices, which has to be done
+  // before enter_runtime
+  if ( !simulated_ ) // only enter the runtime mode once
+  {
+    double tick = Time::get_resolution().get_ms()
+      * kernel().connection_manager.get_min_delay();
+    kernel().music_manager.enter_runtime( tick );
+  }
+}
+
+void
+nest::SimulationManager::simulate( Time const& t )
+{
+  prepare();
+  run( t );
+  cleanup();
+}
+
+void
+nest::SimulationManager::run( Time const& t )
+{
   if ( t == Time::ms( 0.0 ) )
     return;
 
```

Comment:
The next section of code is about error checking. Should that be delegated to a separate method to make the flow of logic clearer?